### PR TITLE
Slamd 95b

### PR DESCRIPTION
**Pull request by StefanLueders**
UI/UX overhaul

### CODE REVIEW

**Review by RugTehseen — COMMENTED** [review]
I added some comments. Most of the changes already look good. Only small things which could be improved

File: tests/discovery/processing/strategies/test_csv_strategy.py [hunk 1]
```diff
@@ -33,3 +59,52 @@ def mock_save_dataset(dataset):
...
+def test_delimiter_parsing_insufficient_rows():
+    headers = 'column1,column2,column3\n'
+    content = ''
+    stream = BytesIO(bytes(headers + content, 'utf-8'))
+    file_data = FileStorage(filename='TestDataset.csv', stream=stream)
+
+    with pytest.raises(SlamdUnprocessableEntityException):
+        CsvStrategy._determine_delimiter(file_data)
+
+
```
Comment (RugTehseen) [inline tests/discovery/processing/strategies/test_csv_strategy.py:94]: There are somehow a lot of empty lines which we do not want to have

File: tests/discovery/processing/strategies/test_csv_strategy.py [hunk 2]
```diff
@@ -20,6 +25,27 @@ def test_create_dataset_parses_file_storage_correctly(monkeypatch):
...
+def test_create_dataset_parsing_error(monkeypatch):
+    headers = 'column1,column2,column3\n'
+    content = '1,2,3\n,6\n7,'
+    stream = BytesIO(bytes(headers + content, 'utf-8'))
+    file_data = FileStorage(filename='temporary.csv', stream=stream)
+
+    with pytest.raises(ValueNotSupportedException):
+        dataset = CsvStrategy.create_dataset(file_data)
+
```
Comment (RugTehseen) [inline tests/discovery/processing/strategies/test_csv_strategy.py:46]: only two empty lines

File: slamd/materials/templates/materials_form.html [hunk 3]
```diff
@@ -1,48 +1,31 @@
-<h3>New material / process</h3>
+<h3 class="mt-3">New material / process</h3>
 <form action="" method="post" novalidate>
     {{ form.csrf_token }}
     <div class="row g-3 mb-3">
         <div class="col-md-6">
             {{ form.material_name.label(class_="control-label", placeholder="Placeholder Text Here") }}
```
[the remark this answers is not in the document]
Comment (StefanLueders) [reply slamd/materials/templates/materials_form.html:6]: The placeholder was assigned to a label, and thus never visible. I removed the property.

File: tests/discovery/processing/strategies/test_csv_strategy.py [hunk 4]
```diff
@@ -20,6 +25,27 @@ def test_create_dataset_parses_file_storage_correctly(monkeypatch):
...
 
 
+def test_create_dataset_filename_error(monkeypatch):
+    headers = 'column1,column2,column3\n'
+    content = '1,2,3\n4,5,6\n7,8,9'
+    stream = BytesIO(bytes(headers + content, 'utf-8'))
+    file_data = FileStorage(filename='temporary.csv', stream=stream)
+
+    with pytest.raises(ValueNotSupportedException):
+        dataset = CsvStrategy.create_dataset(file_data)
```
[the remark this answers is not in the document]
Comment (StefanLueders) [reply tests/discovery/processing/strategies/test_csv_strategy.py:35]: Removed all superfluous local variables.

File: slamd/materials/templates/blending_form.html [hunk 5]
```diff
@@ -1,22 +1,34 @@
...
     {{ blending_name_and_type_form.csrf_token }}
-    <div class="row g-3 mb-3">
-        <div class="col-md-6">
+    <div class="row mb-3">
+        <div class="col-12 col-md-6">
             {{ blending_name_and_type_form.blended_material_name.label(class_="control-label") }}
-            {{ blending_name_and_type_form.blended_material_name(class_="form-control", placeholder="Cannot be empty")
+            {{ blending_name_and_type_form.blended_material_name(class_="form-control mb-3",
+            placeholder="Cannot be empty")
             }}
```
[the remark this answers is not in the document]
Comment (StefanLueders) [reply slamd/materials/templates/blending_form.html:11]: Fixed formatting here and further down the file.

File: slamd/formulations/templates/formulations_min_max_form.html [hunk 6]
```diff
@@ -1,4 +1,4 @@
-{% for entry in formulations_min_max_form.materials_min_max_entries %}
+ {% for entry in formulations_min_max_form.materials_min_max_entries %}
```
[the remark this answers is not in the document]
Comment (StefanLueders) [reply slamd/formulations/templates/formulations_min_max_form.html:1]: Removed unnecessary whitespace

File: slamd/discovery/templates/targets_form.html [hunk 7]
```diff
@@ -1,46 +1,97 @@
...
+    <div class="accordion mb-3">
+        <div class="accordion-item">
+            <h3 class="accordion-header" id="accordionTargetsTable-headingOne">
+                <button class="accordion-button" type="button" data-bs-toggle="collapse"
+                        data-bs-target="#accordionTargetsTable-collapseOne" aria-expanded="true"
+                        aria-controls="accordionTargetsTable-collapseOne">
+                    Show / hide dataset for editing
+                </button>
+            </h3>
+            <div id="accordionTargetsTable-collapseOne" class="accordion-collapse collapse show"
```
[the remark this answers is not in the document]
Comment (StefanLueders) [reply slamd/discovery/templates/targets_form.html:61]: Swapped IDs.

File: slamd/discovery/processing/strategies/csv_strategy.py [hunk 8]
```diff
@@ -1,21 +1,37 @@
...
+            decimal = ','
+
         if file_name == 'temporary.csv':
             raise ValueNotSupportedException('You cannot use the name temporary for your dataset!')
-        return Dataset(name=file_name, dataframe=read_csv(file_data))
+
+        try:
+            dataset = Dataset(name=file_name, dataframe=read_csv(file_data, delimiter=delimiter, decimal=decimal))
+        except:
+            raise ValueNotSupportedException('The dataset you submitted could not be read.')
```
[the remark this answers is not in the document]
Comment (StefanLueders) [reply slamd/discovery/processing/strategies/csv_strategy.py:32]: Previously, the only thing we checked for was the filename (temporary.csv, which we still check for). Here is what the code used to look like:

```
@classmethod
def create_dataset(cls, file_data):
    # Generate a safe filename for the new dataset
    file_name = secure_filename(file_data.filename)

    if file_name == 'temporary.csv':
        raise ValueNotSupportedException('You cannot use the name temporary for your dataset!')
    return Dataset(name=file_name, dataframe=read_csv(file_data))
```

The new try... except is supposed to deal with incorrectly parsed column separators. What other kind of errors should we try to catch?

I don't think there is much we can specifically check for - from the pandas return we can't really tell if there was an issue with the separators, if the file was missing data, or if the user uploaded a binary. All we know is that he uploaded an invalid file.

[the remark this answers is not in the document]
Comment (RugTehseen) [reply slamd/discovery/processing/strategies/csv_strategy.py:32]: ok, I see; makes sense for me now

**Review by RugTehseen — APPROVED** [review]
Now everything looks good

**Pull request merged**